### PR TITLE
[rebalancing] cache: add cache function for looking up containers by cgroup path.

### PR DESCRIPTION
**Pull request by klihub**
Add LookupContainerByCgroup(), a function for finding a container based on a cgroupfs path.
We need this for associating collected cgroupfs based metrics with actual running containers.

### CODE REVIEW

**Comment by mythi** [comment]
TODO: add a test case

File: pkg/cri/resource-manager/cache/cache.go [hunk 1]
```diff
@@ -694,6 +696,35 @@ func (cch *cache) LookupContainer(id string) (Container, bool) {
...
 }
 
+// LookupContainerByCgroup looks up the container for the given cgroup path.
+func (cch *cache) LookupContainerByCgroup(path string) (Container, bool) {
+	cch.Debug("resolving %s to a container...", path)
+
+	for id, c := range cch.Containers {
+		if id != c.CacheID {
+			continue
+		}
```
Comment (marquiz) [inline pkg/cri/resource-manager/cache/cache.go]: When do we hit this, again? Maybe a short comment to clarify(?)

File: pkg/cri/resource-manager/cache/cache.go [hunk 2]
```diff
@@ -694,6 +696,35 @@ func (cch *cache) LookupContainer(id string) (Container, bool) {
...
+		if !strings.HasPrefix(path, parent) {
+			continue
+		}
+
+		if strings.Index(path, c.GetID()) != -1 {
+			return c, true
+		}
+	}
+
+	return nil, false
```
Comment (marquiz) [inline pkg/cri/resource-manager/cache/cache.go]: Would `nil, true` be valid in some case? Could we just return `Container`?

File: pkg/cri/resource-manager/cache/cache.go (hunk 1 above)
Comment (klihub) [reply pkg/cri/resource-manager/cache/cache.go]: We hash containers by two IDs: the ID in the CRI runtime and our cache ID. We hit this for every container which has already been successfully created in the actual runtime. So it is rather the rule, not the exception.

Comment (marquiz) [reply pkg/cri/resource-manager/cache/cache.go]: Ah yeah, of course

File: pkg/cri/resource-manager/cache/cache.go (hunk 2 above)
Comment (klihub) [reply pkg/cri/resource-manager/cache/cache.go]: > Would `nil, true` be valid in some case? Could we just return `Container`?

I think this is the classical go nil interface problem. If we returned (nil, true) here, how could you tell a valid Container apart from a nil Container interface without using reflect ?

Comment (marquiz) [reply pkg/cri/resource-manager/cache/cache.go]: > I think this is the classical go nil interface problem. If we returned (nil, true) here, how could you tell a valid Container apart from a nil Container interface without using reflect ?

Yeah, that's true. What I was thinking was that is it a "legitimate" for these functions to return a `nil` Container interface in any condition?

**Comment by klihub** [comment]
Added minimal tests for cache.LookupContainerByCgroup().

**Comment by klihub** [comment]
> TODO: add a test case

Added (a minimal) test case.

**Pull request merged**